### ISSUE TITLE
False Positive for Object Element Dereference

### ISSUE THREAD

**Issue by Kurt-von-Laven** [rhysd/actionlint#66]
Actionlint knows about array element dereferences, but not object element dereferences. For example, GitHub Actions supports the syntax `${{ join(needs.*.result, ' ') }}` as documented in [slack-templates' README](https://github.com/ScribeMD/slack-templates#summarize-the-results-of-an-entire-workflow) despite the fact that [needs](https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context) is not an array. Actionlint 1.6.6 yields an error message of the following form:

```
[...]/.github/workflows/test.yaml:[row]:[col]: receiver of array element dereference must be type of array but got "{<job_id_1>: {outputs: {}; result: string}; <job_id_2>: {outputs: {}; result: string}}" [expression]
         |
[row] |           results: ${{ join(needs.*.result, ' ') }}
```

Thank you for creating Actionlint! It's quite nice to have a super easy way to check whether an action has syntax errors without wasting build credits.

**Comment by alambike** [comment]
We are seeing the same thing here with this expression: 
```
      - name: Check failures
        if: ${{ contains(steps.*.outcome, 'failure') }}

```
```
.github/workflows/test.yml:128:26: receiver of array element dereference must be type of array but got "{pre: {outcome: string; outputs: {string => string}; conclusion: string}; pro: {outputs: {string => string}; conclusion: string; outcome: string}; metadata: {outputs: {string => string}; conclusion: string; outcome: string}; des: {conclusion: string; outcome: string; outputs: {string => string}}}" [expression]
    |
128 |         if: ${{ contains(steps.*.outcome, 'failure') }}
    |                          ^~~~~~~~~~~~~~~~

```

**Comment by rhysd** [comment]
Thank you for reporting this. It looks a bug of actionlint.

**Comment by rhysd** [comment]
https://docs.github.com/en/actions/learn-github-actions/expressions#object-filters

The document only mentions arrays as target of object filtering. So I thought the syntax was only for arrays when I implemented the type check.

**Comment by Kurt-von-Laven** [comment]
Indeed; I noticed this as well when trying to find a link that would prove the syntax was supported and was pleasantly surprised that it was. I filed the issue linked above with GitHub to enhance their documentation accordingly.

**Comment by rhysd** [comment]
I tried

```yaml
- run: |
    echo '${{ toJSON(fromJSON(env.JSON).*.foo) }}'
  env:
    JSON: '{"a": {"foo": "from a"}, "b": {"foo": "from b"}}'
```

and confirmed output

```
[
  "from a",
  "from b"
]
```

https://github.com/rhysd/actionlint/runs/4136626284?check_suite_focus=true

**Comment by rhysd** [comment]
With cabbf84, `x.*` is deduced to `array<any>` even if `x` is an object. This can prevent the false positive reported in this issue.

**Comment by rhysd** [comment]
The fix was released as v1.6.7

**Comment by Kurt-von-Laven** [comment]
I can confirm that this fix is working for us. Thank you!
